crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,531 @@
+                // Linux VIRTIO SCMI insists on this.
+                // We can accept it and ignore it, the sensor will be still working.
+                Result::Ok(vec![])
+            }
+            SENSOR_READING_GET => {
+                if !self.sensor().enabled {
+                    return Result::Err(ScmiDeviceError::NotEnabled);
+                }
+                self.reading_get()
+            }
+            _ => Result::Err(ScmiDeviceError::UnsupportedRequest),
+        }
+    }
+}
+
+// It's possible to impl ScmiDevice for SensorT but it is not very useful
+// because it doesn't allow to pass SensorT as ScmiDevice directly.
+// Hence this wrapper.
+pub struct SensorDevice(pub(crate) Box<dyn SensorT>);
+
+impl ScmiDevice for SensorDevice {
+    fn initialize(&mut self) -> Result<(), String> {
+        self.0.initialize()
+    }
+
+    fn protocol(&self) -> ProtocolId {
+        self.0.protocol()
+    }
+
+    fn handle(&mut self, message_id: MessageId, parameters: &[MessageValue]) -> DeviceResult {
+        self.0.handle(message_id, parameters)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::assert_eq;
+
+    use super::*;
+
+    #[test]
+    fn test_help() {
+        let help = devices_help();
+        assert!(
+            help.contains("Available devices:\n"),
+            "global label missing"
+        );
+        assert!(help.contains("fake:"), "sensor name missing");
+        assert!(
+            help.contains("fake accelerometer"),
+            "short description missing"
+        );
+        assert!(help.contains("3-axes sensor"), "long description missing");
+        assert!(help.contains("Parameters:\n"), "parameter label missing");
+        assert!(help.contains("- name:"), "parameter `name' missing");
+    }
+
+    fn device_properties() -> DeviceProperties {
+        let s = String::from;
+        DeviceProperties::new(vec![
+            (s("foo"), s("val1")),
+            (s("def"), s("val2")),
+            (s("bar"), s("val3")),
+        ])
+    }
+
+    #[test]
+    fn test_device_properties() {
+        let properties = device_properties();
+        assert_eq!(properties.get("bar"), Some("val3"));
+        assert_eq!(properties.get("baz"), None);
+        assert_eq!(
+            properties.names(),
+            HashSet::from_iter(["foo", "def", "bar"].into_iter())
+        );
+        let expected = ["abc", "def", "ghi"];
+        let missing = properties.missing(&expected);
+        assert_eq!(missing, HashSet::from_iter(["abc", "ghi"].into_iter()));
+        let extra = properties.extra(&expected);
+        assert_eq!(extra, HashSet::from_iter(["foo", "bar"].into_iter()));
+    }
+
+    #[test]
+    fn test_check_device_properties() {
+        let properties = device_properties();
+        let result = check_device_properties(&properties, &["abc", "def", "ghi"], &["foo", "baz"]);
+        assert_eq!(result, Err("Missing device parameters: abc, ghi".into()));
```

Comment:
This sound like the check_device_properties should return a proper Error type instead of a string but no strong opinion.

Reply:
I agree, done.